### PR TITLE
Check for valid PB types in serialize_property

### LINKED ISSUE #2939: Bad error message when passing wrong type to Python resource constructor
For many/most type errors that users might make when passing incorrect values as inputs to resource constructors in Python, they get an error message like:

```
 pulumi:pulumi:Stack (test-dev):
    sys:1: RuntimeWarning: coroutine 'get_instance_profile' was never awaited
 
    error: Program failed with an unhandled exception:
    error: Traceback (most recent call last):
      File "/usr/local/bin/pulumi-language-python-exec", line 85, in <module>
        loop.run_until_complete(coro)
      File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
        return future.result()
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/pulumi/runtime/stack.py", line 38, in run_in_stack
        await RPC_MANAGER.unhandled_exeception()
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/pulumi/runtime/rpc_manager.py", line 68, in rpc_wrapper
        result = await rpc_function(*args, **kwargs)
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/pulumi/runtime/resource.py", line 286, in do_register
        resolver = await prepare_resource(res, ty, custom, props, opts)
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/pulumi/runtime/resource.py", line 79, in prepare_resource
        serialized_props = await rpc.serialize_properties(props, property_dependencies_resources, res.translate_input_property)
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/pulumi/runtime/rpc.py", line 67, in serialize_properties
        struct[translated_name] = result
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/google/protobuf/internal/well_known_types.py", line 765, in __setitem__
        _SetStructValue(self.fields[key], value)
      File "/Users/ablum/gh/pulimi/venv/lib/python3.7/site-packages/google/protobuf/internal/well_known_types.py", line 732, in _SetStructValue
        raise ValueError('Unexpected type')
    ValueError: Unexpected type
    error: an unhandled error occurred: Program exited with non-zero exit code: 1
```

The failure is triggered in protobuf serialization, instead of being caught and reported by the constructor itself.  We need to find some way to report a meaningful error here.

The above can be reproduced with something like the following, where a `Future[Object]`is passed as the argument where a `Input[str]` is expected. 

```
import pulumi
from pulumi_aws import s3, ec2, iam

instance_profile = iam.get_instance_profile(name="webServerProfile-ad563fb")
policy = iam.get_policy(arn="arn:aws:iam::aws:policy/AdministratorAccess")
iam.RolePolicyAttachment(role=instance_profile,resource_name="policy_test",policy_arn=policy)
```

### PR DESCRIPTION
**Pull request by pgavlin**
Just what it says on the tin. This allows us to return an incrementally
better error.

Fixes #2939.

### CODE REVIEW

File: sdk/python/lib/pulumi/runtime/rpc.py [hunk 1]
```diff
@@ -43,6 +44,14 @@
 _special_secret_sig = "1b47061264138c4ac30d75fd1eb44270"
 """special_secret_sig is a randomly assigned hash used to identify secrets in maps. See pkg/resource/properties.go"""
 
+_INT_OR_FLOAT = six.integer_types + (float,)
+
+def isLegalProtobufValue(value: Any) -> bool:
+    """
+    Returns True if the given value is a legal Protobuf value.
+    """
+    return value is None or isinstance(value, (bool, six.string_types, _INT_OR_FLOAT, dict, list))
```
Comment (lukehoban) [inline sdk/python/lib/pulumi/runtime/rpc.py]: Is this documented or discoverable anywhere we can reference here?

File: sdk/python/lib/pulumi/runtime/rpc.py [hunk 2]
```diff
@@ -166,6 +175,10 @@
 
         return obj
 
+    # Ensure that we have a value that Protobuf understands.
+    if not isLegalProtobufValue(value):
+        raise ValueError(f"unexpected input of type {type(value).__name__}")
```
Comment (lukehoban) [inline sdk/python/lib/pulumi/runtime/rpc.py]: Should we be a little more descriptive?  Perhaps “inputs must be primitives, lists or dicts”?

File: sdk/python/lib/test/langhost/resource_op_bad_inputs/test_resource_op_bad_inputs.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# limitations under the License.
+from os import path
+from ..util import LanghostTest
+
+
+class UnhandledExceptionTest(LanghostTest):
+    def test_unhandled_exception(self):
+        self.run_test(
+            program=path.join(self.base_path(), "resource_op_bad_inputs"),
+            expected_error="Program exited with non-zero exit code: 1")
```
Comment (lukehoban) [inline sdk/python/lib/test/langhost/resource_op_bad_inputs/test_resource_op_bad_inputs.py]: I don’t quite follow how this test tests the change here - didn’t we fail before as well (just with a bad error message)?

File: sdk/python/lib/pulumi/runtime/rpc.py (hunk 2 above)
Comment (pgavlin) [reply sdk/python/lib/pulumi/runtime/rpc.py]: Yeah, I wasn't sure how verbose we wanted to be. Do you think it's misleading to omit "Outputs, Awaitables, Resources, Assets, or Archives" from that message? 
Is there a docs page we could point to?

File: sdk/python/lib/test/langhost/resource_op_bad_inputs/test_resource_op_bad_inputs.py (hunk 3 above)
Comment (pgavlin) [reply sdk/python/lib/test/langhost/resource_op_bad_inputs/test_resource_op_bad_inputs.py]: We don't really have a great way to validate the precise error we raise in the integration test framework. Let me see if we have unit tests somewhere.

File: sdk/python/lib/pulumi/runtime/rpc.py (hunk 1 above)
Comment (pgavlin) [reply sdk/python/lib/pulumi/runtime/rpc.py]: We could link to the source: https://github.com/protocolbuffers/protobuf/blob/master/python/google/protobuf/internal/well_known_types.py#L714-L732

As far as I can tell, [the docs](https://developers.google.com/protocol-buffers/docs/reference/python-generated#struct) don't mention this at all.

File: sdk/python/lib/test/langhost/resource_op_bad_inputs/test_resource_op_bad_inputs.py (hunk 3 above)
Comment (pgavlin) [reply sdk/python/lib/test/langhost/resource_op_bad_inputs/test_resource_op_bad_inputs.py]: Added a unit test.

File: sdk/python/lib/pulumi/runtime/rpc.py (hunk 2 above)
Comment (lukehoban) [reply sdk/python/lib/pulumi/runtime/rpc.py]: Ahh - good point that we would need to include these.  I'm okay leaving as you have it for now.  Ideally we would add a docs page to talk about the "Pulumi Object Model" or something like that to describe what the shape of these kinds of values can be.  And then link to that from here once we have it.

File: sdk/python/lib/pulumi/runtime/rpc.py (hunk 1 above)
Comment (lukehoban) [reply sdk/python/lib/pulumi/runtime/rpc.py]: An internal facing comment linking to the source wouldn't hurt.

File: sdk/python/Pipfile [hunk 4]
```diff
@@ -7,6 +7,7 @@ name = "pypi"
 protobuf = ">=3.6.0"
 grpcio = ">=1.9.1"
 dill = ">=0.3.0"
+six = ">=1.12.0"
```
Comment (lukehoban) [inline sdk/python/Pipfile]: do we strictly need `six`?  I would have thought that was only needed for Python2 compatibility?

Comment (pgavlin) [reply sdk/python/Pipfile]: I think you're correct. I used it to match `protobuf`, so I'd like to keep it as-is if that's okay.

File: sdk/python/lib/pulumi/runtime/rpc.py (hunk 1 above)
Comment (pgavlin) [reply sdk/python/lib/pulumi/runtime/rpc.py]: Done

**Pull request merged**